code/modules/mob/living/carbon/xenomorph/castes/wraith/abilities_wraith.dm

```diff
@@ -0,0 +1,460 @@
+		for(var/turf/T2 in turfs)
+			if(get_dist(owner, T2) == WRAITH_BLINK_RANGE) //Find the turf at the max distance we can go and break
+				T = T2
+				break
+			else
+				continue
+
+	if(!can_use_ability(T)) //Since we updated the turf, check it again.
+		fail_activate()
+
+	var/mob/living/carbon/xenomorph/wraith/TI = owner
+	var/mob/pulled
+
+	if(ismob(TI.pulling)) //bring the target we're pulling with us, but at the cost of sharply increasing the next cooldown
+		pulled = TI.pulling
+
+	blink_warp(TI, T, pulled)
+
+	succeed_activate()
+	add_cooldown()
+
+	GLOB.round_statistics.wraith_blinks++
+	SSblackbox.record_feedback("tally", "round_statistics", 1, "wraith_blinks") //Statistics
+
+
+/datum/action/xeno_action/activable/blink/proc/blink_warp(mob/living/carbon/xenomorph/wraith/TI, turf/T, mob/pulled = null) //This handles the location swap between the Wraith and the Warp Shadow
+
+	TI.face_atom(T) //Face the target so we don't look like an ass
+
+	teleport_debuff_aoe(TI) //Debuff when we vanish
+
+	TI.forceMove(T) //Teleport to our target turf
+
+	if(pulled) //bring the pulled target with us if applicable
+		cooldown_timer *= WRAITH_BLINK_DRAG_MULTIPLIER
+		to_chat(TI, "<span class='xenodanger'>We bring [pulled.name] with us. We won't be ready to blink again for [cooldown_timer * 0.1] seconds due to the strain of doing so.</span>")
+
+	teleport_debuff_aoe(TI) //Debuff when we reappear
+
+
+/datum/action/xeno_action/proc/teleport_debuff_aoe(atom/movable/teleporter, silent = FALSE, no_visuals = FALSE) //Called by many of the Wraith' teleportation effects
+
+	if(!teleporter) //Sanity
+		return
+
+	var/mob/living/carbon/xenomorph/source = owner
+
+	if(!silent) //Sound effects
+		playsound(teleporter, 'sound/effects/EMPulse.ogg', 25, 1) //Sound at the location we are arriving at
+
+	if(!no_visuals) //Visuals
+		new /obj/effect/temp_visual/blink_portal(get_turf(teleporter))
+
+	for(var/turf/affected_tile in range(1,teleporter.loc))
+		affected_tile.Shake(4, 4, 1 SECONDS) //SFX
+		for(var/mob/living/target in affected_tile)
+			target.Shake(4, 4, 1 SECONDS) //SFX
+			if(target.stat == DEAD)
+				continue
+			if(isxeno(target))
+				var/mob/living/carbon/xenomorph/X = target
+				if(X.hive == source.hive) //No friendly fire
+					continue
+
+			shake_camera(target, 2, 1)
+			target.adjust_stagger(WRAITH_TELEPORT_DEBUFF_STACKS)
+			target.add_slowdown(WRAITH_TELEPORT_DEBUFF_STACKS)
+			target.adjust_drugginess(WRAITH_TELEPORT_DEBUFF_STACKS) //minor visual distortion
+			to_chat(target, WRAITH_TELEPORT_DEBUFF_MSG)
+
+/datum/action/xeno_action/evasion/on_cooldown_finish()
+	if(!owner) //Sanity
+		return
+
+	if(cooldown_timer > initial(cooldown_timer) ) //Reset the cooldown if increased from dragging someone.
+		to_chat(owner, "<span class='xenonotice'>We are able to take blink again.</span>")
+		owner.playsound_local(owner, 'sound/effects/xeno_newlarva.ogg', 25, 0, 1)
+		cooldown_timer = initial(cooldown_timer)
+	return ..()
```

Comment:
What the hell is this even

Reply:
Blink's next cooldown is increased significantly as a balancing mechanism if used to transport a target we're dragging. This cooldown proc normalizes the cooldown duration.